subsys/bluetooth/controller/flash_driver.c

```diff
@@ -0,0 +1,313 @@
+	LOG_DBG("Flash op %u complete", flash_state.op);
+	if (flash_state.op == FLASH_OP_WRITE) {
+		LOG_HEXDUMP_DBG(flash_state.data, flash_state.prev_len, "wrote");
+	}
+
+	int err;
+	flash_state.addr += flash_state.prev_len;
+	flash_state.data = (const void *) ((intptr_t) flash_state.data + flash_state.prev_len);
+	flash_state.len -= flash_state.prev_len;
+
+	if (flash_state.len > 0) {
+		err = flash_op_execute();
+		/* All inputs should have been validated on the first call. */
+		__ASSERT(err == 0, "Continued flash operation failed\n");
+	} else {
+		flash_state.op = FLASH_OP_NONE;
+		k_sem_give(&flash_state.sem);
+	}
+}
+
+static size_t offset_32(const void *addr)
+{
+	return (const u32_t) addr & 0x3;
+}
+
+static size_t unaligned_word_copy(u32_t *word_dst, void *dst, const void *src, size_t len)
+{
+	/* Example: Writing one byte from src to dst via word_dst.
+	 *
+	 * |    |<----- src (len = 1)
+	 * |  |<------- dst
+	 * | 0 1 2 3 |
+	 * | address |
+	 *
+	 * max_offset             : max(1, 2) => 2
+	 * remaining_bytes_in_word: 4 - 2     => 2
+	 * bytes_to_copy          : min(1, 2) => 1
+	 */
+	size_t max_offset = max(offset_32(dst), offset_32(src));
+	size_t remaining_bytes_in_word = sizeof(u32_t) - max_offset;
+	size_t bytes_to_copy = min(len, remaining_bytes_in_word);
+
+	/* nRF52832 Product specification:
+	 *  Only full 32-bit words can be written to Flash using the
+	 *  NVMC interface. To write less than 32 bits to Flash, write
+	 *  the data as a word, and set all the bits that should remain
+	 *  unchanged in the word to '1'. */
+	*word_dst = ~0;
+	memcpy(&((u8_t *)word_dst)[offset_32(dst)], src, bytes_to_copy);
+	return bytes_to_copy;
+}
+
+static int flash_op_write(void)
+{
+	if (is_aligned_32(flash_state.addr) &&
+	    is_aligned_32((off_t) flash_state.data) &&
+	    flash_state.len >= sizeof(u32_t)) {
+		flash_state.prev_len = min(align_32(flash_state.len), NRF_FICR->CODEPAGESIZE);
+		return ble_controller_flash_write((u32_t) flash_state.addr,
+						  flash_state.data,
+						  bytes_to_words(flash_state.prev_len),
+						  flash_operation_complete_callback);
+	} else {
+		flash_state.prev_len = unaligned_word_copy(&flash_state.tmp_word,
+							   (void *)flash_state.addr,
+							   flash_state.data,
+							   flash_state.len);
+		return ble_controller_flash_write((u32_t) align_32(flash_state.addr),
+						  &flash_state.tmp_word,
+						  1,
+						  flash_operation_complete_callback);
+	}
+}
+
+static int flash_op_execute(void)
+{
+	int err;
+
+	err = MULTITHREADING_LOCK_ACQUIRE();
+	if (!err) {
+		if (flash_state.op == FLASH_OP_WRITE) {
+			err = flash_op_write();
+		} else if (flash_state.op == FLASH_OP_ERASE) {
+			flash_state.prev_len = NRF_FICR->CODEPAGESIZE;
+			err = ble_controller_flash_page_erase((u32_t) flash_state.addr,
+							      flash_operation_complete_callback);
+		} else {
+			__ASSERT(0, "Unsupported operation\n");
+			err = -EINVAL;
+		}
+	}
+	MULTITHREADING_LOCK_RELEASE();
```

Comment:
Seems like MULTITHREADING_LOCK_RELEASE(); will be called even if MULTITHREADING_LOCK_ACQUIRE(); fails to acquire the lock.

Reply:
:+1: I will move it up